core/light-client/src/concurrent_access.rs

```diff
@@ -0,0 +1,145 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+//! Concurrent access mechanisms that ensure mutually exclusive read/write access
+//! to the light-client (validator) by employing RwLocks under the hood.
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::{
+	error::{Error, Result},
+	LightClientState, Validator,
+};
+use finality_grandpa::BlockNumberOps;
+use itp_sgx_io::SealedIO;
+use lazy_static::lazy_static;
+use sp_runtime::traits::{Block as BlockT, NumberFor};
+use std::marker::PhantomData;
+
+lazy_static! {
+	// As long as we have a file backend, we use this 'dummy' lock,
+	// which guards against concurrent read/write access.
+	pub static ref VALIDATOR_LOCK: RwLock<()> = Default::default();
+}
+
+/// Retrieve an exclusive lock on a validator for either read or write access.
+///
+/// In order to hide the whole locks mechanics, we provide an interface that allows executing
+/// either a mutating, or a non-mutating function on the validator.
+/// The reason we have this additional wrapper around `SealedIO`, is that we need
+/// to guard against concurrent access by using RWLocks (which `SealedIO` does not do).
+pub trait ValidatorAccess<PB>
+where
+	PB: BlockT,
+	NumberFor<PB>: BlockNumberOps,
+{
+	type ValidatorType: Validator<PB> + LightClientState<PB>;
+
+	fn execute_on_validator<F, R>(&self, getter_function: F) -> Result<R>
+	where
+		F: FnOnce(&Self::ValidatorType) -> Result<R>;
+
+	fn execute_mut_on_validator<F, R>(&self, mutating_function: F) -> Result<R>
+	where
+		F: FnOnce(&mut Self::ValidatorType) -> Result<R>;
+}
```

Comment:
All new code in here. A new trait for accessing the validator/light-client. By using the pattern of passing in a function to execute, we can completely hide the `RwLock`s that are used underneath.